File: 프로그래머스/3/150366. 표 병합/표 병합.py
def dfs(stack, visited, word, unmerge=0):
    while stack:
        num = stack.pop()
        if visited[num]:
            continue
        else:
            visited[num] = 1
            
        graph[num] = word
        
        for i in merge[num]:
            if visited[i]:
                continue
            stack.append(i)
        
        if unmerge:
            merge[num] = []

def doMerge(cell1, cell2): # cell2를 cell1으로 바꿈
    visited = [0] * 2500
    stack = [cell2]
    content = graph[cell1]
    dfs(stack, visited, content)
        
def doUnmerge(cell):
    visited = [0] * 2500
    visited[cell] = 1
    stack = []
    
    for i in merge[cell]:
        stack.append(i)
    
    merge[cell] = []
    dfs(stack, visited, "", 1)
        
def Update(cell, word):
    graph[cell] = word
    visited = [0] * 2500
    visited[cell] = 1
    stack = []
    
    for i in merge[cell]:
        stack.append(i)
    
    dfs(stack, visited, word)

graph = [""] * 2500
merge = [[] for _ in range(2500)]
        
def solution(commands):
    answer = []
    
    for i in commands:
        command = i.split(" ")
        
        if command[0] == "UPDATE" and len(command) == 4:
            r, c = int(command[1])-1, int(command[2])-1
            cell = r*50+c
            Update(cell, command[3])
        
            
        elif command[0] == "UPDATE" and len(command) == 3:
            for i in range(2500):
                if graph[i] == command[1]:
                    graph[i] = command[2]
            
        elif command[0] == "MERGE":
            r1, c1, r2, c2 = map(int, command[1:])
            cell1 = (r1-1) * 50 + c1-1
            cell2 = (r2-1) * 50 + c2-1
            
            if cell1 == cell2 : # r, c가 동일하면 스킵
                continue
                
            elif graph[cell1] != "" and graph[cell2] == "":  # 셀1 만 값을 가질 경우
                doMerge(cell1, cell2)
            
            elif graph[cell1] == "" and graph[cell2] != "": # 셀2 만 값을 가질 경우
                doMerge(cell2, cell1)
                
            else: # 둘다 있거나 없다면 셀1값으로 가짐
                doMerge(cell1, cell2)
                
            merge[cell1].append(cell2)
            merge[cell2].append(cell1)
            
        elif command[0] == "UNMERGE":
            r, c = int(command[1])-1, int(command[2])-1
            cell = r * 50 + c
            doUnmerge(cell)
                
        elif command[0] == "PRINT":
            r, c = int(command[1])-1, int(command[2])-1
            word = graph[r * 50 + c]
            if word == "":
                answer.append("EMPTY")
            else:
                answer.append(word)
        
    return answer

File: 프로그래머스/3/150366. 표 병합/표 병합.java
import java.util.*;

public class Solution {
    public static String[] graph = new String[2500];
    public static List<Integer>[] merge = new ArrayList[2500];

    public static List<String> solution(String[] commands) {
        
        for (int i = 0; i < 2500; i++) {
            graph[i] = "";
            merge[i] = new ArrayList<>();
        }
        
        List<String> answer = new ArrayList<>();

        for (String cmd : commands) {
            String[] command = cmd.split(" ");

            if (command[0].equals("UPDATE") && command.length == 4) {
                int r = Integer.parseInt(command[1]) - 1;
                int c = Integer.parseInt(command[2]) - 1;
                int cell = r * 50 + c;
                update(cell, command[3]);

            } else if (command[0].equals("UPDATE") && command.length == 3) {
                for (int i = 0; i < 2500; i++) {
                    if (graph[i].equals(command[1])) {
                        graph[i] = command[2];
                    }
                }

            } else if (command[0].equals("MERGE")) {
                int r1 = Integer.parseInt(command[1]) - 1;
                int c1 = Integer.parseInt(command[2]) - 1;
                int r2 = Integer.parseInt(command[3]) - 1;
                int c2 = Integer.parseInt(command[4]) - 1;

                int cell1 = r1 * 50 + c1;
                int cell2 = r2 * 50 + c2;

                if (cell1 == cell2) {
                    continue;
                } else if (!graph[cell1].isEmpty() && graph[cell2].isEmpty()) {
                    doMerge(cell1, cell2);
                } else if (graph[cell1].isEmpty() && !graph[cell2].isEmpty()) {
                    doMerge(cell2, cell1);
                } else {
                    doMerge(cell1, cell2);
                }

                merge[cell1].add(cell2);
                merge[cell2].add(cell1);

            } else if (command[0].equals("UNMERGE")) {
                int r = Integer.parseInt(command[1]) - 1;
                int c = Integer.parseInt(command[2]) - 1;
                int cell = r * 50 + c;
                doUnmerge(cell);

            } else if (command[0].equals("PRINT")) {
                int r = Integer.parseInt(command[1]) - 1;
                int c = Integer.parseInt(command[2]) - 1;
                String word = graph[r * 50 + c];

                if (word.isEmpty()) {
                    answer.add("EMPTY");
                } else {
                    answer.add(word);
                }
            }
        }

        return answer;
    }
    
    public static void dfs(ArrayDeque<Integer> stack, boolean[] visited, String word) {
        dfs(stack, visited, word, false);
    }
    
    public static void dfs(ArrayDeque<Integer> stack, boolean[] visited, String word, boolean unMerge) {
        while (stack.size() > 0) {
            int cur = stack.pop();
            
            if (visited[cur] == true) {
                continue;
            } else{
                visited[cur] = true;
            }
            
            graph[cur] = word;
            
            for (int i : merge[cur]) {
                if (visited[i] == true) {
                    continue;
                }
                stack.push(i);
            }
            
            if (unMerge == true) {
                merge[cur].clear();
            }
        }
    }

    public static void update(int cell, String value) {
        boolean[] visited = new boolean[2500];
        graph[cell] = value;
        visited[cell] = true;
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        
        for (int i : merge[cell]) {
            stack.push(i);
        }
        
        dfs(stack, visited, value);
        
    }

    public static void doMerge(int cell1, int cell2) {
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        boolean[] visited = new boolean[2500];
        visited[cell2] = true;
        graph[cell2] = graph[cell1];
        
        for (int i : merge[cell2]){
            stack.push(i);
        }
        
        dfs(stack, visited, graph[cell1]);
        
    }

    public static void doUnmerge(int cell) {
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        boolean[] visited = new boolean[2500];
        visited[cell] = true;
        for (int i : merge[cell]) {
            stack.push(i);
        }
        merge[cell].clear();
        dfs(stack, visited, "", true);
    }
}